napari/components/_tests/test_layer_slicer.py

```
@@ -244,3 +244,32 @@ def test_slice_layers_async_task_to_layers_lock(layer_slicer):
 
     assert task.result()[layer].id == 1
     assert task not in layer_slicer._layers_to_task
+
+
+def test_slice_layers_exception_main_thread(layer_slicer):
+    class FakeAsyncLayerError(FakeAsyncLayer):
+        """Exception is raised on the main thread from an error on the main
+        thread immediately when the task is created."""
+
+        def _make_slice_request(self, dims) -> FakeSliceRequest:
+            assert current_thread() == main_thread()
+            raise RuntimeError('Purposeful error.')
+
+    layer = FakeAsyncLayerError()
+    with pytest.raises(RuntimeError):
+        layer_slicer.slice_layers_async(layers=[layer], dims=Dims())
+
+
+def test_slice_layers_exception_subthread_on_result(layer_slicer):
+    """Exception is raised on the main thread from an error on a subthread
+    only after result is called, not upon submission of the task."""
+
+    class FakeAsyncLayerError(FakeAsyncLayer):
+        def _get_slice(self, request: FakeSliceRequest) -> FakeSliceResponse:
+            assert current_thread() != main_thread()
+            raise RuntimeError('Purposeful error.')
+
+    layer = FakeAsyncLayerError()
+    future = layer_slicer.slice_layers_async(layers=[layer], dims=Dims())
+    with pytest.raises(RuntimeError):
```

In theory, we don't control the order execution here - i.e. line 273 could raise before we get to line 274. In practice, that's unlikely, but I'd like to see some more control of execution here as in the other tests.

---

Oh wait, maybe that's the point? Since the executor will swallow the exception until we call `result`? That's not entirely clear - would be great if there's a way to wait until we know the task is done, but before `result` is called.

---

I think you're right, its safer if we add in some control. I've added in a wait to ensure that the future is `done` before calling result. Do you think that covers it well enough?